trending_data/src/main/webapp/script.js

```diff
@@ -1,34 +1,153 @@
+
 let map;
+let windowsHandler;
 
 /* eslint-disable no-unused-vars */
 /**
-
  * Initialise map
  */
 function initMap() {
-  const initPos = new google.maps.LatLng(-34.397, 150.644);
+  const initPos = new google.maps.LatLng(0, 0);
   map = new google.maps.Map(document.getElementById('map'), {
     center: initPos,
-    zoom: 8,
+    zoom: 3,
+    minZoom: 2,
   });
+  windowsHandler = new UniqueWindowHandler(map);
+  addAllMarkers(map);
+}
 
-  const infoString =
-    '<h2>This is a hardcoded tweet</h2>' +
-    '<p>@twitter_person says: Super popular #tweet :o</p>';
+/**
+ * Adds one marker for each country to the map
+ * See CountryCodeServlet.java for more details
+ * @param {Object} map
+ */
+function addAllMarkers(map) {
+  fetch('/countries').then((response) => response.json())
+      .then((countries) => {
+        for (const country of countries) {
+          addMarkerToMapGivenCountry(country, map);
+        }
+      });
+}
 
-  const infoWindow = new google.maps.InfoWindow({
-    content: infoString,
-    maxWidth: 200,
-  });
+/**
+ * Adds one marker for the given country
+ * See CountryCodeServlet.java for more details
+ * @param {Object} country A JSON object {name: String,
+ * alpha2Code: String, lng: number, lat: number}
+ * @param {Object} map
+ */
+function addMarkerToMapGivenCountry(country, map) {
+  addMarkerToMapGivenInfo(country.name, country.alpha2Code,
+      country.lat, country.lng, map);
+}
 
+/**
+ * Adds one marker given country name, code, average longitude and latitude
+ * @param {string} countryName
+ * @param {string} countryCode  (ISO 3166-1) Alpha-2 code
+ * @param {number} lat Latitude (average)
+ * @param {number} lng Longitude (average)
+ * @param {Object} map
+ */
+function addMarkerToMapGivenInfo(countryName, countryCode, lat, lng, map) {
   const marker = new google.maps.Marker({
-    position: initPos,
-    map,
-    title: 'Trending Tweet',
+    position: {lat, lng},
+    map: map,
+    title: countryName,
   });
+  marker.countryCode = countryCode;
 
+  /* For each new marker, listen for a click event; If marker is clicked =>
+  fetch posts for the country corresponding to that marker and display them */
   marker.addListener('click', () => {
-    infoWindow.open(map, marker);
```

Comment:
Maybe displayPosts() can only be called if a window isn't currently open. That would reduce the number of useless API calls. So something like:
```
marker.addListener('click', () => {
  if (!windowsHandler.isInfoWindowOpen()) {
    displayPosts(marker);
  }
}
```
Not sure if this would work but I think worth trying.

Reply:
Follow up: This isn't possible unless we change the implementation.
(not a priority)